File: LCAofBT.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

//Time complexity : O(n)
//Space complexity : O(n)

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == nullptr || root == p || root == q)
            return root;
           
        TreeNode* left = lowestCommonAncestor(root->left, p , q);
        TreeNode* right = lowestCommonAncestor(root->right, p , q);
        
        if(left == nullptr && right == nullptr)
            return nullptr;
        else if(left == nullptr && right != nullptr)
            return right;
        else if(left != nullptr && right == nullptr)
            return left;
        else 
            return root;
    }
};

File: LCA of BT.py
# // Time Complexity : O(n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No

class Solution:
    
    def __init__(self):
        # to store LCA node
        self.ans = None
        
    def lowestCommonAncestor(self, root, p,q ):
        
        def recurse(cur):
            if not cur:
                return False
            
            left = recurse(cur.left)
            
            right = recurse(cur.right)
            
            # if cur is p or q
            mid = cur == p or cur == q
            
            # if two or more flags are true
            if mid + left + right >= 2:
                self.ans = cur
                
            return mid or left or right
        
        recurse(root)
        return self.ans